Document `HoeffdingTree`

rcurtin opened the pull request:
Here's rendered documentation: https://www.ratml.org/misc/mlpack-markdown-doc/hoeffding_tree.html

As with a lot of the previous examples, I had to make fairly intrusive changes to `Train()` to match the API required by the hyperparameter tuner.  Because we don't yet require C++17 (where I could use `std::optional`), there are a lot of overloads.  Once we do, the code can be cleaned up.

The only other functionality I added was a user-facing `Reset()` function to reset the tree, plus a few accessors to get properties of the trees.

I also added a number of tests for everything that I implemented, as well as a test to make sure that training on floating-point data works.

shrit commented:
@rcurtin it is ready to be reviewed ?

shrit commented:
Perfect, will give it a chance tomorrow

shrit commented on src/mlpack/methods/hoeffding_trees/hoeffding_tree.hpp:
```diff
@@ -194,44 +265,136 @@ class HoeffdingTree
+             const data::DatasetInfo& info,
+             const arma::Row<size_t>& labels,
+             const size_t numClasses,
+             const bool batchTraining,
+             const double successProbability);
```
Can we make the `sucessProbability` as ElemType?

rcurtin replied:
I didn't do this, mostly so that when you switch to using a different `ElemType`, as few things change in the API as possible.  (That will also make the documentation job in #3584 easier.)  For something like the success probability, using a `double` or a `float` or whatever other element type really won't make any serious difference, so I think it might be easier to leave as-is.

shrit commented on src/mlpack/methods/hoeffding_trees/hoeffding_tree.hpp:
```diff
@@ -194,44 +265,136 @@ class HoeffdingTree
+             const arma::Row<size_t>& labels,
+             const size_t numClasses,
+             const bool batchTraining,
+             const double successProbability,
```
Same as above

shrit commented on src/mlpack/methods/hoeffding_trees/hoeffding_tree.hpp:
```diff
@@ -194,44 +265,136 @@ class HoeffdingTree
-             const size_t numClasses = 0);
+             const size_t numClasses,
+             const bool batchTraining,
+             const double successProbability,
```
Same as above

shrit reviewed (changes requested):
Very good, similar comments to the last review of the logistic regression class the following links are not working:
![Screenshot 2023-12-15 at 14-20-04 https __www ratml org](https://github.com/mlpack/mlpack/assets/15039782/31734753-a0ac-45fe-b887-62b73ba94906)

Also regarding the fitness function, it would be great to have more details about them if we have many, because currently, we are only mentioning the Gini Impurity.
I am not an expert in trees, so I have no idea how many fitness functions you can have, or what we have already implemented. 

![Screenshot 2023-12-15 at 18-10-26 https __www ratml org](https://github.com/mlpack/mlpack/assets/15039782/8ea24d53-3f28-4042-8eb0-771798a435ed)

The last thing is `double` about the precision, but apart from that great work, we can merge this really soon

rcurtin commented:
> Very good, similar comments to the last review of the logistic regression class the following links are not working

Yep, the invalid links will be handled in a follow-up PR. :+1: 

> Also regarding the fitness function, it would be great to have more details about them if we have many, because currently, we are only mentioning the Gini Impurity. I am not an expert in trees, so I have no idea how many fitness functions you can have, or what we have already implemented.

Good point.  I added a sentence or two with some links in b74b8bc476; let me know what you think. :+1:

> The last thing is `double` about the precision, but apart from that great work, we can merge this really soon

In this case, actually the way `HoeffdingTree` is written internally, the fitness functions should always return `double` regardless of the `ElemType` of the original data.  This is because the fitness functions for the Hoeffding tree use an `arma::Mat<size_t>` containing class counts.  Like I mentioned in another comment, I don't think it'll make any measurable difference to use `double`, `float`, or whatever other type there, so probably just best to leave it as `double` in my opinion.
